internal/surfer/gcloud/builder.go

```diff
@@ -0,0 +1,403 @@
+			fullCollectionPath := fmt.Sprintf("%s.%s", shortServiceName, baseCollectionPath)
+
+			// We assemble and return the `ResourceSpec`.
+			return &ResourceSpec{
+				Name:                  name,
+				PluralName:            pluralName,
+				Collection:            fullCollectionPath,
+				DisableAutoCompleters: true,
+				Attributes:            newAttributesFromSegments(segments),
+			}, nil
+		}
+	}
+	return nil, fmt.Errorf("resource definition not found for type %q", field.ResourceReference.Type)
+}
+
+// newAttributesFromSegments parses a structured resource pattern and extracts the attributes
+// that make up the resource's name.
+func newAttributesFromSegments(segments []api.PathSegment) []Attribute {
+	var attributes []Attribute
+
+	// We iterate over the segments of the pattern.
+	for i, part := range segments {
+		// A variable segment is enclosed in curly braces.
+		if part.Variable != nil {
+			// The `attribute_name` is the name of the variable (e.g., "project").
+			if len(part.Variable.FieldPath) == 0 {
+				continue
+			}
+			name := part.Variable.FieldPath[len(part.Variable.FieldPath)-1]
+			var parameterName string
+
+			// The `parameter_name` is derived from the preceding literal segment
+			// (e.g., "projects" -> "projectsId"). This is a gcloud convention.
+			if i > 0 && segments[i-1].Literal != nil {
+				parameterName = *segments[i-1].Literal + "Id"
+			} else {
+				// This is a fallback for the unlikely case that a pattern starts with a variable.
+				parameterName = name + "sId"
+			}
+
+			attr := Attribute{
+				AttributeName: name,
+				ParameterName: parameterName,
+				Help:          fmt.Sprintf("The %s id of the {resource} resource.", name),
+			}
+
+			// If the attribute is a project, we add the standard gcloud property fallback,
+			// so users don't have to specify `--project` if it's already configured.
+			if name == "project" {
+				attr.Property = "core/project"
+			}
+			attributes = append(attributes, attr)
+		}
+	}
+	return attributes
+}
+
+// newRequest creates the `Request` part of the command definition.
+func newRequest(method *api.Method, overrides *Config, model *api.API) *Request {
+	// TODO(https://github.com/googleapis/librarian/issues/3290): The collection path is partially hardcoded.
+	return &Request{
+		APIVersion: apiVersion(overrides),
+		Collection: []string{fmt.Sprintf("parallelstore.projects.locations.%s", utils.GetPluralResourceNameForMethod(method, model))},
+	}
+}
+
+// newAsync creates the `Async` part of the command definition for long-running operations.
+func newAsync(_ *api.Method, _ *Config) *Async {
+	return &Async{
+		// TODO(https://github.com/googleapis/librarian/issues/3290): The collection path is partially hardcoded.
+		Collection: []string{"parallelstore.projects.locations.operations"},
+	}
+}
+
+// findHelpTextRule finds the help text rule from the config that applies to the current method.
+func findHelpTextRule(method *api.Method, overrides *Config) *HelpTextRule {
```

Comment:
Nit: I feel like this does not belong in this file. There should be an abstraction of the config that has these helpers

Reply:
Yes, I agree that this clutters the builder. I do intend to enhance this function eventually, but I haven't been able to dedicate significant time to it yet. I've added a todo with a related issue I had created a while ago. 

https://github.com/googleapis/librarian/issues/3033